certora/specs/Tokens.spec

```
@@ -0,0 +1,137 @@
+    // Safe require because it is a verified invariant in Morpho Blue.
+    require lastUpdated > 0 => Morpho.libId(marketParams) == id;
+
+    return marketParams;
+}
+
+function summarySupply(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) returns (uint256, uint256) {
+    assert shares == 0;
+    assert onBehalf == currentContract;
+    assert data.length == 0;
+
+    // Safe require because it is a verified invariant.
+    require hasSupplyCapIsEnabled(Morpho.libId(marketParams));
+    // Safe require because it is a verified invariant.
+    require isEnabledHasConsistentAsset(marketParams);
+
+    // Summarize supply as just a transfer for the purpose of this specification file, which is sound because only the properties about tokens are verified in this file.
+    Util.safeTransferFrom(marketParams.loanToken, currentContract, MORPHO(), assets);
+
+    return (assets, shares);
+}
+
+function summaryWithdraw(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) returns (uint256, uint256) {
+    assert onBehalf == currentContract;
+    assert receiver == currentContract;
+
+    MetaMorphoHarness.Id id = Morpho.libId(marketParams);
+    uint256 withdrawn = Util.withdrawnAssets(MORPHO(), id, assets, shares);
+
+    bool enabled;
+    _, enabled, _ = config(id);
+    // Safe require because:
+    // - for reallocate this is checked in the code
+    // - for withdraw, it is verified that markets in the withdraw queue are enabled.
+    require enabled;
+    // Safe require because it is a verified invariant.
+    require isEnabledHasConsistentAsset(marketParams);
+
+    // Summarize supply as just a transfer for the purpose of this specification file, which is sound because only the properties about tokens are verified in this file.
+    Util.safeTransferFrom(marketParams.loanToken, MORPHO(), currentContract, withdrawn);
+
+    return (withdrawn, shares);
+}
+
+// Check that, on deposit, MetaMorpho's balance does not change and that Morpho's balance increases by the corresponding amount.
+rule depositTokenChange(env e, uint256 assets, address receiver) {
+    address asset = asset();
+    address morpho = MORPHO();
+
+    // Trick to require that all the following addresses are different.
+    require morpho == 0x10;
+    require asset == 0x11;
+    require currentContract == 0x12;
+    require e.msg.sender == 0x13;
+    require receiver == 0x14;
+
+    uint256 balanceMorphoBefore = Util.balanceOf(asset, morpho);
+    uint256 balanceMetaMorphoBefore = Util.balanceOf(asset, currentContract);
+    deposit(e, assets, receiver);
+    uint256 balanceMorphoAfter = Util.balanceOf(asset, morpho);
+    uint256 balanceMetaMorphoAfter = Util.balanceOf(asset, currentContract);
+
+    assert assert_uint256(balanceMorphoAfter - balanceMorphoBefore) == assets;
+    assert balanceMetaMorphoAfter == balanceMetaMorphoBefore;
```

```suggestion
    uint256 balanceMorphoBefore = Util.balanceOf(asset, morpho);
    uint256 balanceMetaMorphoBefore = Util.balanceOf(asset, currentContract);
    uint256 balanceSenderBefore = Util.balanceOf(asset, e.msg.sender);
    deposit(e, assets, receiver);
    uint256 balanceMorphoAfter = Util.balanceOf(asset, morpho);
    uint256 balanceMetaMorphoAfter = Util.balanceOf(asset, currentContract);
    uint256 balanceSenderAfter = Util.balanceOf(asset, e.msg.sender);

    assert assert_uint256(balanceMorphoAfter - balanceMorphoBefore) == assets;
    assert balanceMetaMorphoAfter == balanceMetaMorphoBefore;
    assert assert_uint256(balanceSenderBefore - balanceSenderAfter) == assets;
```

I imagine that you thought about it. Why doesn't it work?

---

I need to investigate this more

---

Added, it seems to work fine actually